### PR TITLE
Dt-863 abuse test cases implementation.

### PR DESCRIPTION
**Pull request by preetamnpr**

### CODE REVIEW

**Comment by preetamnpr** [comment]
Merge to release branch until we have the final decision on releases of betas.

File: booking/src/main/java/org/dcsa/conformance/standards/booking/BookingScenarioListBuilder.java [hunk 1]
```diff
@@ -78,101 +78,131 @@ private BookingScenarioListBuilder thenAllPathsFrom(BookingState bookingState) {
   private BookingScenarioListBuilder thenAllPathsFrom(
       BookingState bookingState, BookingState originalBookingState) {
     return switch (bookingState) {
-      case CANCELLED, COMPLETED, DECLINED, REJECTED  -> then(shipper_GetBooking(bookingState));
+      case COMPLETED  -> then(
+        shipper_GetBooking(bookingState).thenEither(
+         uc7_shipper_submitBookingAmendment(true)
+            .then(shipper_GetBooking(bookingState))
+        )
```
Comment (nt-gt) [inline booking/src/main/java/org/dcsa/conformance/standards/booking/BookingScenarioListBuilder.java:85]: Any reason why `cancel` is not being tried here as well?

File: booking/src/main/java/org/dcsa/conformance/standards/booking/BookingScenarioListBuilder.java [hunk 2]
```diff
@@ -78,101 +78,131 @@ private BookingScenarioListBuilder thenAllPathsFrom(BookingState bookingState) {
...
+      );
+      case DECLINED -> then(
+        shipper_GetBooking(bookingState).then(
+          uc7_shipper_submitBookingAmendment(true)
+            .then(shipper_GetBooking(bookingState))
+        )
```
Comment (nt-gt) [inline booking/src/main/java/org/dcsa/conformance/standards/booking/BookingScenarioListBuilder.java:91]: Any reason why `cancel` is not being tried here as well?

File: booking/src/main/java/org/dcsa/conformance/standards/booking/BookingScenarioListBuilder.java [hunk 3]
```diff
@@ -78,101 +78,131 @@ private BookingScenarioListBuilder thenAllPathsFrom(BookingState bookingState) {
...
+      );
+      case CANCELLED  -> then(
+        shipper_GetBooking(bookingState).thenEither(
+          uc3_shipper_submitUpdatedBookingRequest(true)
+            .then(shipper_GetBooking(bookingState))
+        )
```
Comment (nt-gt) [inline booking/src/main/java/org/dcsa/conformance/standards/booking/BookingScenarioListBuilder.java:97]: If `cancel` is not explicitly allowed on top of `cancel` then we should definitely have `cancel` here as well (for the return code check)

File: booking/src/main/java/org/dcsa/conformance/standards/booking/BookingScenarioListBuilder.java [hunk 4]
```diff
@@ -78,101 +78,131 @@ private BookingScenarioListBuilder thenAllPathsFrom(BookingState bookingState) {
   private BookingScenarioListBuilder thenAllPathsFrom(
       BookingState bookingState, BookingState originalBookingState) {
     return switch (bookingState) {
-      case CANCELLED, COMPLETED, DECLINED, REJECTED  -> then(shipper_GetBooking(bookingState));
+      case COMPLETED  -> then(
+        shipper_GetBooking(bookingState).thenEither(
+         uc7_shipper_submitBookingAmendment(true)
```
Comment (nt-gt) [inline booking/src/main/java/org/dcsa/conformance/standards/booking/BookingScenarioListBuilder.java:83]: General remark: Please rename and use a different name for abuse test cases when they are used in the state transitions.

In the eBL, I named them `auc_X` (for `AbUse Case`).

this makes it easier to see at a distance that these transitions are invalid.


Note: I am fine with them being funneled into the same underling code (action classes) after that. The point is that I want to quickly be able to tell "valid" from "invalid" transition apart something as complex as this. A `true` / `false` will not cut it as `true` / `false` are non-descriptive (especially when the editor is not there to provide inlay hints, which it is not in GitHub among other)

File: booking/src/main/java/org/dcsa/conformance/standards/booking/BookingScenarioListBuilder.java [hunk 5]
```diff
@@ -78,101 +78,131 @@ private BookingScenarioListBuilder thenAllPathsFrom(BookingState bookingState) {
...
+            .then(shipper_GetBooking(bookingState))
```
Comment (nt-gt) [inline booking/src/main/java/org/dcsa/conformance/standards/booking/BookingScenarioListBuilder.java:84]: Do we not check the state in `shipper_GetBooking`? For eBL the expected status is passed to the GET call.

Also, for an amendment call, you should also check that the amended booking is unchanged

File: booking/src/main/java/org/dcsa/conformance/standards/booking/BookingScenarioListBuilder.java (hunk 1 above)
Comment (preetamnpr) [reply booking/src/main/java/org/dcsa/conformance/standards/booking/BookingScenarioListBuilder.java:85]: As per the story user performs UC7 when Booking status is COMPLETED and it hasn't been defined in the story.

File: booking/src/main/java/org/dcsa/conformance/standards/booking/BookingScenarioListBuilder.java (hunk 2 above)
Comment (preetamnpr) [reply booking/src/main/java/org/dcsa/conformance/standards/booking/BookingScenarioListBuilder.java:91]: As per the story [DT-863](https://dcsa.atlassian.net/browse/DT-863) user performs UC7 when Booking status is DECLINED and it hasn't been defined in the story.

File: booking/src/main/java/org/dcsa/conformance/standards/booking/BookingScenarioListBuilder.java (hunk 1 above)
Comment (nt-gt) [reply booking/src/main/java/org/dcsa/conformance/standards/booking/BookingScenarioListBuilder.java:85]: Challenge the story then to confirm whether this omission is intentional or an oversight

File: booking/src/main/java/org/dcsa/conformance/standards/booking/BookingScenarioListBuilder.java (hunk 2 above)
Comment (nt-gt) [reply booking/src/main/java/org/dcsa/conformance/standards/booking/BookingScenarioListBuilder.java:91]: Challenge the story then to confirm whether this omission is intentional or an oversight

File: booking/src/main/java/org/dcsa/conformance/standards/booking/BookingScenarioListBuilder.java [hunk 6]
```diff
@@ -78,7 +80,24 @@ private BookingScenarioListBuilder thenAllPathsFrom(BookingState bookingState) {
...
+      case DECLINED, CANCELLED, REJECTED -> then(
+        shipper_GetBooking(bookingState).thenEither(
+          auc_shipper_sendInvalidBookingAction(UPDATE_BOOKING)
+            .then(shipper_GetBooking(bookingState)),
+          auc_shipper_sendInvalidBookingAction(SUBMIT_BOOKING_AMENDMENT)
+            .then(shipper_GetBooking(bookingState)),
+          auc_shipper_sendInvalidBookingAction(CANCEL_BOOKING)
+            .then(shipper_GetBooking(bookingState))
+        )
+      );
```
Comment (nt-gt) [inline booking/src/main/java/org/dcsa/conformance/standards/booking/BookingScenarioListBuilder.java:100]: Missing a happy path flow (where there is no `auc` step)

File: booking/src/main/java/org/dcsa/conformance/standards/booking/BookingScenarioListBuilder.java [hunk 7]
```diff
@@ -109,16 +128,15 @@ private BookingScenarioListBuilder thenAllPathsFrom(
...
+                uc3_shipper_submitUpdatedBookingRequest()
+                    .thenAllPathsFrom(UPDATE_RECEIVED),
+                uc4_carrier_rejectBookingRequest().thenAllPathsFrom(REJECTED),
+                uc5_carrier_confirmBookingRequest().thenAllPathsFrom(CONFIRMED),
+                uc11_shipper_cancelBooking().thenAllPathsFrom(CANCELLED)));
       case START -> thenEither(
-        uc1_shipper_SubmitBookingRequest(ScenarioType.REGULAR).thenAllPathsFrom(RECEIVED),
-        uc1_shipper_SubmitBookingRequest(ScenarioType.REEFER).thenHappyPathFrom(RECEIVED),
-        uc1_shipper_SubmitBookingRequest(ScenarioType.DG).thenHappyPathFrom(RECEIVED));
+        uc1_shipper_SubmitBookingRequest().thenAllPathsFrom(RECEIVED));
```
Comment (nt-gt) [inline booking/src/main/java/org/dcsa/conformance/standards/booking/BookingScenarioListBuilder.java:139]: I do not understand what made you remove the other scenario types?

Comment (preetamnpr) [reply booking/src/main/java/org/dcsa/conformance/standards/booking/BookingScenarioListBuilder.java:139]: Good catch.. I was testing on local machine with only regular scenario hence I had to delete it.. will roll it back.

Comment (preetamnpr) [reply booking/src/main/java/org/dcsa/conformance/standards/booking/BookingScenarioListBuilder.java:139]: I could recall now, those functions were redundant and not being used any more. Hence removed them.

File: booking/src/main/java/org/dcsa/conformance/standards/booking/BookingScenarioListBuilder.java (hunk 6 above)
Comment (preetamnpr) [reply booking/src/main/java/org/dcsa/conformance/standards/booking/BookingScenarioListBuilder.java]: done..

**Review by nt-gt — APPROVED** [review]

**Pull request merged**